srcts/src/components/darkModeSwitch.ts

```diff
@@ -0,0 +1,298 @@
+      }
+
+      svg line,
+      svg circle {
+        vector-effect: non-scaling-stroke;
+      }
+    `,
+  ];
+
+  // eslint-disable-next-line @typescript-eslint/no-empty-function, @typescript-eslint/no-unused-vars
+  onChangeCallback: (x: boolean) => void = (x: boolean) => {};
+
+  // onValueChange = makeValueChangeEmitter(this, this.id);
+
+  connectedCallback(): void {
+    super.connectedCallback();
+
+    this.themeAttribute =
+      this.getAttribute("theme-attribute") || "data-shinytheme";
+
+    if (window.matchMedia("(prefers-color-scheme: dark)").matches) {
+      this.themeValue = "dark";
+    }
+
+    this.setPreference();
+
+    // Sync with system changes
+    window
+      .matchMedia("(prefers-color-scheme: dark)")
+      .addEventListener("change", ({ matches: isDark }) => {
+        this.themeValue = isDark ? "dark" : "light";
+        this.setPreference();
+      });
+  }
+
+  getValue(): string {
+    return this.themeValue;
+  }
+
+  render(): ReturnType<LitElement["render"]> {
+    return html`
+      <button
+        title="Switch to ${this.themeValue === "light" ? "dark" : "light"} mode"
+        aria-label="auto"
+        aria-live="polite"
+        data-theme="${this.themeValue}"
+        @click="${this.onClick}"
+      >
+        <svg class="sun-and-moon" aria-hidden="true" viewBox="0 0 24 24">
+          <mask class="moon" id="moon-mask">
+            <rect x="0" y="0" width="100%" height="100%" fill="white" />
+            <circle cx="25" cy="10" r="6" fill="black" />
+          </mask>
+          <circle
+            class="sun"
+            cx="12"
+            cy="12"
+            r="6"
+            mask="url(#moon-mask)"
+            fill="currentColor"
+          />
+          <g class="sun-beams" stroke="currentColor">
+            <line x1="12" y1="1" x2="12" y2="3" />
+            <line x1="12" y1="21" x2="12" y2="23" />
+            <line x1="4.22" y1="4.22" x2="5.64" y2="5.64" />
+            <line x1="18.36" y1="18.36" x2="19.78" y2="19.78" />
+            <line x1="1" y1="12" x2="3" y2="12" />
+            <line x1="21" y1="12" x2="23" y2="12" />
+            <line x1="4.22" y1="19.78" x2="5.64" y2="18.36" />
+            <line x1="18.36" y1="5.64" x2="19.78" y2="4.22" />
+          </g>
+        </svg>
+      </button>
+    `;
+  }
+
+  receiveMessage(el: DarkModeSwitch, data: ToggleMessage): void {
+    if (data.method === "toggle") {
+      if (data.value === "toggle") {
+        data.value = this.themeValue === "light" ? "dark" : "light";
+      }
+      el.setAttribute("theme-value", data.value);
+    }
+  }
+
+  onClick(e: MouseEvent): void {
+    e.stopPropagation();
+    this.themeValue = this.themeValue === "light" ? "dark" : "light";
+    this.setPreference();
+    this.onChangeCallback(true);
+  }
+
+  updated(changedProperties: Map<string, any>): void {
+    if (changedProperties.has("themeValue")) {
+      this.setPreference();
+      this.onChangeCallback(true);
+    }
+  }
```

Comment:
I'm not sure if it's better or actually matters, but it seems, at least from the name, that `attributeChangedCallback()` is designed for this?

Reply:
If we had to make decisions about what to do with the old or new value of the attribute, then `attributeChangedCallback()` would be needed. We're letting lit handle reflecting the property value change back to the web component attribute, so I think `updated()` might be slightly more ergonomic.

We can even simplify the logic a bit and have all theme changes handled by `updated()`, so everywhere else we only need to set the attribute value. ff5d96d43b06a7c3db965e0b920767fbeaa8e338